crates/iota-genesis-builder/src/stake.rs

```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+        if timestamp > 0 {
+            staked_with_timelock.push((surplus_nanos, timestamp));
+        }
+        // If the `surplus_object` is returned by `maybe_reuse_surplus`, then it means
+        // it used all its `surplus_nanos` and it can be burned.
+        if let Some(surplus_object) = surplus_object_option {
+            to_burn.push(surplus_object);
+        }
+        // Else, if the `surplus_object` was not completely drained, then we
+        // don't need to continue. In this case `allocation_tot_amount ==
+        // target_amount`.
     }
-}
 
-/// Create the necessary allocations to cover `amount_nanos` for all
-/// `validators`.
-///
-/// This function iterates in turn over [`TimeLock`] and
-/// [`GasCoin`][iota_types::gas_coin::GasCoin] objects created
-/// during stardust migration that are owned by the `delegator`.
-pub fn delegate_genesis_stake<'info>(
-    validators: impl Iterator<Item = &'info GenesisValidatorInfo>,
-    delegator: IotaAddress,
-    migration_objects: &MigrationObjects,
-    amount_nanos: u64,
-) -> anyhow::Result<GenesisStake> {
-    let timelocks_pool = migration_objects.get_sorted_timelocks_and_expiration_by_owner(delegator);
-    let gas_coins_pool = migration_objects.get_gas_coins_by_owner(delegator);
-    if timelocks_pool.is_none() && gas_coins_pool.is_none() {
-        anyhow::bail!("no timelocks or gas-coin objects found for delegator {delegator:?}");
+    if allocation_tot_amount < target_amount {
+        to_burn.append(
+            &mut pool
+                .by_ref()
+                .map_while(|(object, timestamp)| {
+                    if allocation_tot_amount < target_amount {
+                        let difference_from_target = target_amount - allocation_tot_amount;
+                        let obj_ref = object.compute_object_reference();
+                        let object_balance = get_gas_balance_maybe(object)?.value();
+
+                        if object_balance <= difference_from_target {
+                            if timestamp > 0 {
+                                staked_with_timelock.push((object_balance, timestamp));
+                            }
+                            allocation_tot_amount += object_balance;
+                            // Place `obj_ref` in `to_burn` and continue
+                            Some(obj_ref)
+                        } else {
+                            surplus = CoinSurplus {
+                                obj_ref: Some(obj_ref),
+                                surplus_nanos: object_balance - difference_from_target,
+                                timestamp,
+                            };
+                            if timestamp > 0 {
+                                staked_with_timelock.push((difference_from_target, timestamp));
+                            }
+                            allocation_tot_amount += difference_from_target;
+                            // Do NOT place `obj_ref` in `to_burn` and break
```

Comment:
```suggestion
                            // Do NOT place `obj_ref` in `to_burn` because it is reused in the CoinSurplus and then break the map_while
```